### PR TITLE
FS: Add FS messaging socket

### LINKED ISSUE #32: IN: Messaging sockets
Integrate sockets into messaging so that messages are sent instantly when users are online

### PR DESCRIPTION
**Pull request by navjot-chahal**
### What this PR does:
- Adds complete instant messaging feature with socket.io.
- Shows users online in the conversation sidebar.
- Removes/Fixes any previous issues with BE socket.io event handlers.

**_Note: This PR also contains features/code from the previous PR which are pending approval._**

### Screenshots / Videos:
- Shown below is the video demonstration of the feature added.

https://user-images.githubusercontent.com/60498899/139451688-2e62597e-d087-467f-b5a9-51b75d4d24c4.mov

### Any information needed to test this feature:
- Perform `npm install` in both server and client folders.
- The feature also requires a pre-populated DB with conversations setup (can be set up manually with Postman).

### Any issues with the current functionality:
- The ability to start a new conversation is dependent on the request acceptance feature as per the discussion in a team meeting. Upon accepting a request a conversation can be generated with a simple API call.
- Conversation starting ability can be easily added in here later on if required.

closes #32

### CODE REVIEW

File: client/src/components/Search/Search.tsx [hunk 1]
```diff
@@ -65,7 +65,7 @@ const Search = ({ search, handleChange }: Props): JSX.Element => {
           setOpen(false);
         }}
         getOptionSelected={(option, value) => option.username === value.username}
-        getOptionLabel={(option) => option.username}
+        // getOptionLabel={(option) => option?.username}
```
Comment (sundayezeilo) [inline client/src/components/Search/Search.tsx:68]: Let's remove unused code

File: server/app.js [hunk 2]
```diff
@@ -28,11 +30,50 @@ const server = http.createServer(app);
...
+io.use(socketAuthVerify);
+
 io.on("connection", (socket) => {
-  console.log("connected");
+  if (!onlineUserLog.checkInLog(socket.user)) {
+    const newUser = onlineUserLog.addUser(socket.user, socket.id);
+    if (!newUser) return;
+
+    const allUsersOnline = onlineUserLog.getAll();
+    socket.broadcast.emit("allUsersOnlineRes", allUsersOnline);
```
Comment (sundayezeilo) [inline server/app.js:45]: :+1:

File: server/utils/onlineUserLog.js [hunk 3]
```diff
@@ -0,0 +1,37 @@
+class onlineUserLog {
+  constructor() {
+    this.log = new Object();
```
Comment (sundayezeilo) [inline server/utils/onlineUserLog.js]: Nice use of object for logging online users. This makes the access time constant, ie, `O(1)`

**Review by sundayezeilo — COMMENTED** [review]
This is good! Just a minor comment below.
Happy coding!

File: client/src/components/Search/Search.tsx (hunk 1 above)
Comment (navjot-chahal) [reply client/src/components/Search/Search.tsx:68]: @ezeilo-su Sure, this was not my scope of work for this issue. Must have been by mistake, I will fix it. Thanks.

File: client/src/components/Messages/ChatSideBanner/ChatSideBanner.tsx [hunk 4]
```diff
@@ -0,0 +1,36 @@
+import { Box, Typography } from '@material-ui/core';
+import useStyles from './useStyles';
+import SideBannerItem from '../sideBannerItem/SideBannerItem';
+import { useConversation } from '../../../context/useConversationContext';
+import { useSocket } from '../../../context/useSocketContext';
+import { useEffect, useState } from 'react';
+
+const ChatSideBanner = (): JSX.Element => {
```
Comment (rajivtitus) [inline client/src/components/Messages/ChatSideBanner/ChatSideBanner.tsx:8]: A lot of this code seems to be from a previous PR, please adjust the base of the PR or rebase to only reflect the changes :)

Comment (rajivtitus) [reply client/src/components/Messages/ChatSideBanner/ChatSideBanner.tsx:8]: @navjot-chahal please see comment above

**Comment by navjot-chahal** [comment]
@rajivtitus The PR commit is cleaned now to only reflect changes related to the #32 issue.

**Comment by navjot-chahal** [comment]
@ezeilo-su @ugur-akin Can I get a "go ahead" on this if everything else is fine?

**Review by rajivtitus — APPROVED** [review]
Lgtm! 👍

**Pull request merged**